### ISSUE TITLE
CHAPTER 13 Data API: Putting Everything together (sub-section)

### ISSUE THREAD

**Issue by kasrahabib** [ageron/handson-ml2#407]
Hi @ageron,
This question might be very simple; but somehow I still manage to not get it.
I know what does the ```repeat()``` do. 
In the second line of the following code you chained ```repeat()``` function, for ```train_set``` you set it to ```None```, which I did not understand the logic behind it; is it because the data is not sufficient to train and we want to generate enough of it. If that is the case then training on the exact duplicate of the instances does not make sense, does it?

I played with code and removed ```repeat()```. If removed then the model only executes for 2 epochs and throws ```Input run out of data``` error.

Thanks in advance!

``` python

def csv_reader_dataset(filepaths, repeat=1, n_readers=5,
                       n_read_threads=None, shuffle_buffer_size=10000,
                       n_parse_threads=5, batch_size=32):
    dataset = tf.data.Dataset.list_files(filepaths, seed = 42).repeat(repeat)
    dataset = dataset.interleave(
        lambda filepath: tf.data.TextLineDataset(filepath).skip(1),
        cycle_length = n_readers, num_parallel_calls = n_read_threads)
    
    dataset = dataset.shuffle(shuffle_buffer_size)
    dataset = dataset.map(preprocess, num_parallel_calls = n_parse_threads)
    dataset = dataset.batch(batch_size)
    return dataset.prefetch(1)

train_set = csv_reader_dataset(train_filepaths, repeat = None)
valid_set = csv_reader_dataset(valid_filepaths)
test_set = csv_reader_dataset(test_filepaths)


keras.backend.clear_session()
np.random.seed(42)
tf.random.set_seed(42)

model = keras.models.Sequential([
    keras.layers.InputLayer(input_shape = X_train.shape[-1: ]),
    keras.layers.Dense(30, activation = 'relu'),
    keras.layers.Dense(1)
])

m_loss = keras.losses.mean_squared_error
m_optimizer = keras.optimizers.SGD(lr = 1e-3)

batch_size = 32
model.compile(loss = m_loss, optimizer = m_optimizer, metrics = ['accuracy'])
model.fit(train_set, steps_per_epoch = len(X_train) // batch_size, epochs = 10, validation_data = valid_set)
```

**Comment by ageron** [comment]
Hi @kasrahabib,

Thanks for your question. When we set `repeat=None`, it means that the dataset will be repeated forever. For example, try running this code:

```python
import tensorflow as tf

for x in tf.data.Dataset.range(10).repeat(3):
    print(x)
```

This will print tensors from 0 to 9 three times. But if you replace `3` with `None` (or just remove `3`), it will keep printing tensors from 0 to 9 forever.

Coming back to the book's code, using `repeat(None)` for `train_set` means that the training set will be an infinitely repeating dataset. Since we don't training to last forever, we must specify the number of `steps_per_epoch` and the number of `epochs` when calling the `fit()` method. This way, the `fit()` method will know when to stop reading samples from the infinitely repeating training set.

A simpler way to do this nowadays is to avoid repeating the training set (don't call `repeat()` at all), then avoid using `steps_per_epoch`. Indeed Keras will train the model perfectly well: it will just go through the entire dataset at each epoch.

The reason why I used the first approach in the book is that there used to be a bug in Keras in TF 2.0 which made it impossible to use the simpler approach: you just had to use the first approach. It later became possible to use the second approach, but it did not correctly handle dataset shuffling: the dataset was shuffled just once at the beginning of training, instead of once per epoch. So I preferred to stick with the first approach: a bit uglier, but at least it worked fine. However, these issues have been solved now, so you can just remove `repeat()` and `steps_per_epoch`, and everything will work just fine.

**Comment by kasrahabib** [comment]
Hi again,  @ageron

I was thinking about it for almost 3 days; did a lot of googling but could not find something that satisfy me.
Thanks for the kind reply!
It all makes sense now.


Tons of love from Afghanistan! ☺️

**Comment by ageron** [comment]
Hi @kasrahabib ,
I'm glad I could help! 😊
It's so great to get feedback from all over the world!
